File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.java
import java.util.*;

class Solution {
    public int solution(int[][] jobs) {
        
        Arrays.sort(jobs, (a, b) -> a[0] - b[0]);
        
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[1] - b[1]);
        
        int time = 0;
        int answer = 0;
        int index = 0;
        int finish = 0;
        
        while(finish < jobs.length){
            while(index < jobs.length && time >= jobs[index][0]){
                pq.add(jobs[index++]);
            }
            
            if(pq.isEmpty()){
                time = jobs[index][0];
            }else{
                int out[] = pq.poll();
                time += out[1];
                answer += time - out[0];
                finish++;
            }
        }
        
       
        return answer/jobs.length;
    }
}

File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.js
function solution(jobs) {
    
    jobs.sort((a,b) => a[0] - b[0]);
    
    var que = [];
    var time = 0;
    var index = 0;
    var answer = 0;
    var finish = 0;
    
    while(finish < jobs.length){
        
        while(index < jobs.length && time >= jobs[index][0]){
            que.push(jobs[index++]);
        }
        
        que.sort((a,b) => b[1] - a[1]);
        
        if(!que.length){
            time = jobs[index][0];
        }else{
            var out = que.pop();
            time += out[1];
            answer += time - out[0];
            finish++;
        }
        
    }
    
    return Math.floor(answer/jobs.length);
}